packages/core/saved-objects/core-saved-objects-server/src/saved_objects_type.ts

```diff
@@ -128,6 +132,102 @@ export interface SavedObjectsType<Attributes = any> {
    * An optional {@link SavedObjectsTypeManagementDefinition | saved objects management section} definition for the type.
    */
   management?: SavedObjectsTypeManagementDefinition<Attributes>;
+
+  /**
+   * A map of model versions associated with this type.
+   *
+   * Model versions supersede the {@link SavedObjectsType.migrations | migrations} (and {@link SavedObjectsType.schemas | schemas}) APIs
+   * by exposing an unified way of describing the changes of shape or data of the type.
+   *
+   * Model versioning is decoupled from Kibana versioning, and isolated between types.
+   * Model versions are identified by a single numeric value, starting at `1` and without gaps.
+   *
+   * Please refer to {@link SavedObjectsModelVersion} for details on the API's usages.
+   *
+   * A **valid** versioning would be:
+   *
+   * ```ts
+   * {
+   *   name: 'foo',
+   *   // other mandatory attributes...
+   *   modelVersions: {
+   *     '1': modelVersion1,
+   *     '2': modelVersion2,
+   *     '3': modelVersion3,
+   *   }
+   * }
+   * ```
+   *
+   * A **invalid** versioning would be:
+   *
+   * ```ts
+   * {
+   *   name: 'foo',
+   *   // other mandatory attributes...
+   *   modelVersions: {
+   *     '1': modelVersion1,
+   *     '3': modelVersion3, // ERROR, no model version 2
+   *     '3.1': modelVersion31, // ERROR, model version is a single numeric value
+   *   }
+   * }
+   * ```
+   *
+   * @alpha experimental and subject to change.
+   */
+  modelVersions?: SavedObjectsModelVersionMap | SavedObjectsModelVersionMapProvider;
```

Comment:
The modelVersion API, as exposed on the `SavedObjectsType` type. 

Are we fine with using a map based definition as we did for the old migration registration? It's kinda inelegant given versions are only integers, but  I couldn't find a better idea.

We could imagine using a list with the version number being included in the entries, but I wasn't convinced if was necessarily better, especially given internally we'll likely have to convert it to a map anyway?

Reply:
A map is fine. It may not be elegant but it works for now.